File: sdks/clojure/src/mongo_do/collection.clj
(ns mongo-do.collection
  "MongoDB collection operations with data-oriented API.

   All operations work with plain Clojure maps.
   Query results can be consumed as lazy sequences."
  (:require [mongo-do.types :as types]
            [mongo-do.cursor :as cursor]))

;; ============================================================================
;; Collection Protocol
;; ============================================================================

(defprotocol ICollection
  "Protocol for MongoDB collection operations."
  (-name [this] "Get the collection name.")
  (-full-name [this] "Get the full collection name (db.collection).")
  (-database [this] "Get the parent database.")
  (-rpc [this] "Get the RPC client.")
  (-db-name [this] "Get the database name."))

;; ============================================================================
;; Collection Implementation
;; ============================================================================

(defrecord Collection [rpc database coll-name]
  ICollection
  (-name [_] coll-name)

  (-full-name [_]
    (str (.-db-name database) "." coll-name))

  (-database [_] database)

  (-rpc [_] rpc)

  (-db-name [_] (.-db-name database)))

;; ============================================================================
;; Projection Helpers
;; ============================================================================

(defn- normalize-projection
  "Normalize projection to a map.
   Accepts either a map or a vector of field names."
  [projection]
  (cond
    (nil? projection) nil
    (map? projection) projection
    (sequential? projection) (into {} (map (fn [f] [f 1]) projection))
    :else nil))

;; ============================================================================
;; Insert Operations
;; ============================================================================

(defn insert-one!
  "Insert a single document.

   Args:
     coll - Collection instance
     document - Document map to insert

   Returns:
     InsertOneResult map with :inserted-id and :acknowledged

   Throws:
     ExceptionInfo with :type :duplicate-key if document with same _id exists

   Example:
     (insert-one! users {:name \"Alice\" :email \"alice@example.com\"})"
  [coll document]
  (let [doc (types/ensure-id document)
        rpc (-rpc coll)
        db-name (-db-name coll)
        coll-name (-name coll)
        result (-> rpc .-mongo (.insertOne db-name coll-name doc))]
    (types/parse-insert-one-result result (:_id doc))))

(defn insert-many!
  "Insert multiple documents.

   Args:
     coll - Collection instance
     documents - Vector of document maps to insert
     opts - Optional configuration:
            :ordered - If true, stop on first error (default: true)

   Returns:
     InsertManyResult map with :inserted-ids and :acknowledged

   Example:
     (insert-many! users [{:name \"Alice\"} {:name \"Bob\"}])"
  ([coll documents]
   (insert-many! coll documents {}))
  ([coll documents opts]
   (let [docs (mapv types/ensure-id documents)
         rpc (-rpc coll)
         db-name (-db-name coll)
         coll-name (-name coll)
         result (-> rpc .-mongo (.insertMany db-name coll-name docs
                                             {:ordered (get opts :ordered true)}))]
     (types/parse-insert-many-result result (mapv :_id docs)))))

;; ============================================================================
;; Find Operations
;; ============================================================================

(defn find-one
  "Find a single document.

   Args:
     coll - Collection instance
     filter - Query filter map (optional, default: {})
     opts - Optional configuration:
            :projection - Fields to include/exclude (map or vector)

   Returns:
     Document map, or nil if not found.

   Example:
     (find-one users {:email \"alice@example.com\"})
     (find-one users {:status \"active\"} {:projection [:name :email]})"
  ([coll]
   (find-one coll {} nil))
  ([coll filter]
   (find-one coll filter nil))
  ([coll filter opts]
   (let [rpc (-rpc coll)
         db-name (-db-name coll)
         coll-name (-name coll)
         options (cond-> {}
                   (:projection opts)
                   (assoc :projection (normalize-projection (:projection opts))))
         result (-> rpc .-mongo (.findOne db-name coll-name (or filter {}) options))]
     (when (and result
                (map? result)
                (not (:error result)))
       result))))

(defn find
  "Find documents matching the filter.

   Args:
     coll - Collection instance
     filter - Query filter map (optional, default: {})
     opts - Optional configuration:
            :projection - Fields to include/exclude
            :sort - Sort specification [[field dir] ...]
            :skip - Number of documents to skip
            :limit - Maximum documents to return
            :batch-size - Cursor batch size

   Returns:
     Vector of document maps.

   Example:
     (find users {:status \"active\"})
     (find users {:age {:$gt 21}} {:sort [[:name 1]] :limit 10})"
  ([coll]
   (find coll {} nil))
  ([coll filter]
   (find coll filter nil))
  ([coll filter opts]
   (let [rpc (-rpc coll)
         db-name (-db-name coll)
         coll-name (-name coll)
         options (cond-> {}
                   (:projection opts)
                   (assoc :projection (normalize-projection (:projection opts)))
                   (:sort opts)
                   (assoc :sort (:sort opts))
                   (:skip opts)
                   (assoc :skip (:skip opts))
                   (:limit opts)
                   (assoc :limit (:limit opts)))
         result (-> rpc .-mongo (.find db-name coll-name (or filter {}) options))]
     (if (sequential? result)
       (vec result)
       []))))

(defn find-seq
  "Find documents as a lazy sequence.

   Args:
     coll - Collection instance
     filter - Query filter map (optional, default: {})
     opts - Optional configuration (same as find)

   Returns:
     Lazy sequence of document maps.

   Example:
     (doseq [user (find-seq users {:status \"active\"})]
       (println (:name user)))

     ;; With transducers
     (into []
           (comp (filter #(> (:age %) 21))
                 (map :email)
                 (take 100))
           (find-seq users {}))"
  ([coll]
   (find-seq coll {} nil))
  ([coll filter]
   (find-seq coll filter nil))
  ([coll filter opts]
   (cursor/->lazy-seq (-rpc coll)
                      (-db-name coll)
                      (-name coll)
                      (or filter {})
                      opts)))

;; ============================================================================
;; Update Operations
;; ============================================================================

(defn update-one!
  "Update a single document.

   Args:
     coll - Collection instance
     filter - Query filter map
     update - Update operations map (e.g., {:$set {:status \"active\"}})
     opts - Optional configuration:
            :upsert - If true, insert if no document matches (default: false)

   Returns:
     UpdateResult map with :matched-count, :modified-count, :upserted-id

   Example:
     (update-one! users {:email \"alice@example.com\"}
                        {:$set {:status \"vip\"}})"
  ([coll filter update]
   (update-one! coll filter update {}))
  ([coll filter update opts]
   (let [rpc (-rpc coll)
         db-name (-db-name coll)
         coll-name (-name coll)
         result (-> rpc .-mongo (.updateOne db-name coll-name filter update
                                            {:upsert (get opts :upsert false)}))]
     (types/parse-update-result result))))

(defn update-many!
  "Update multiple documents.

   Args:
     coll - Collection instance
     filter - Query filter map
     update - Update operations map
     opts - Optional configuration:
            :upsert - If true, insert if no document matches (default: false)

   Returns:
     UpdateResult map with :matched-count, :modified-count

   Example:
     (update-many! users {:status \"inactive\"}
                         {:$set {:archived true}})"
  ([coll filter update]
   (update-many! coll filter update {}))
  ([coll filter update opts]
   (let [rpc (-rpc coll)
         db-name (-db-name coll)
         coll-name (-name coll)
         result (-> rpc .-mongo (.updateMany db-name coll-name filter update
                                             {:upsert (get opts :upsert false)}))]
     (types/parse-update-result result))))

(defn replace-one!
  "Replace a single document.

   Args:
     coll - Collection instance
     filter - Query filter map
     replacement - Replacement document
     opts - Optional configuration:
            :upsert - If true, insert if no document matches (default: false)

   Returns:
     UpdateResult map

   Example:
     (replace-one! users {:_id \"123\"}
                         {:_id \"123\" :name \"Alice\" :status \"vip\"})"
  ([coll filter replacement]
   (replace-one! coll filter replacement {}))
  ([coll filter replacement opts]
   (let [rpc (-rpc coll)
         db-name (-db-name coll)
         coll-name (-name coll)
         result (-> rpc .-mongo (.replaceOne db-name coll-name filter replacement
                                             {:upsert (get opts :upsert false)}))]
     (types/parse-update-result result))))

;; ============================================================================
;; Delete Operations
;; ============================================================================

(defn delete-one!
  "Delete a single document.

   Args:
     coll - Collection instance
     filter - Query filter map

   Returns:
     DeleteResult map with :deleted-count

   Example:
     (delete-one! users {:email \"alice@example.com\"})"
  [coll filter]
  (let [rpc (-rpc coll)
        db-name (-db-name coll)
        coll-name (-name coll)
        result (-> rpc .-mongo (.deleteOne db-name coll-name filter))]
    (types/parse-delete-result result)))

(defn delete-many!
  "Delete multiple documents.

   Args:
     coll - Collection instance
     filter - Query filter map

   Returns:
     DeleteResult map with :deleted-count

   Example:
     (delete-many! users {:status \"inactive\"})"
  [coll filter]
  (let [rpc (-rpc coll)
        db-name (-db-name coll)
        coll-name (-name coll)
        result (-> rpc .-mongo (.deleteMany db-name coll-name filter))]
    (types/parse-delete-result result)))

;; ============================================================================
;; Count and Distinct
;; ============================================================================

(defn count-documents
  "Count documents matching the filter.

   Args:
     coll - Collection instance
     filter - Query filter map (optional, default: {})

   Returns:
     Number of matching documents.

   Example:
     (count-documents users {:status \"active\"})"
  ([coll]
   (count-documents coll {}))
  ([coll filter]
   (let [rpc (-rpc coll)
         db-name (-db-name coll)
         coll-name (-name coll)
         result (-> rpc .-mongo (.countDocuments db-name coll-name (or filter {})))]
     (if (integer? result) result 0))))

(defn estimated-document-count
  "Get an estimated count of documents in the collection.

   This is faster than count-documents but may not be accurate.

   Returns:
     Estimated number of documents.

   Example:
     (estimated-document-count users)"
  [coll]
  (let [rpc (-rpc coll)
        db-name (-db-name coll)
        coll-name (-name coll)
        result (-> rpc .-mongo (.estimatedDocumentCount db-name coll-name))]
    (if (integer? result) result 0)))

(defn distinct-values
  "Get distinct values for a field.

   Args:
     coll - Collection instance
     key - Field name to get distinct values for
     filter - Query filter map (optional, default: {})

   Returns:
     Vector of distinct values.

   Example:
     (distinct-values users :status)
     (distinct-values orders :category {:customer-id \"123\"})"
  ([coll key]
   (distinct-values coll key {}))
  ([coll key filter]
   (let [rpc (-rpc coll)
         db-name (-db-name coll)
         coll-name (-name coll)
         result (-> rpc .-mongo (.distinct db-name coll-name (name key) (or filter {})))]
     (if (sequential? result) (vec result) []))))

;; ============================================================================
;; Aggregation
;; ============================================================================

(defn aggregate
  "Run an aggregation pipeline.

   Args:
     coll - Collection instance
     pipeline - Vector of aggregation stages

   Returns:
     Vector of aggregation results.

   Example:
     (aggregate users
       [{:$match {:status \"active\"}}
        {:$group {:_id \"$country\" :count {:$sum 1}}}
        {:$sort {:count -1}}])"
  [coll pipeline]
  (let [rpc (-rpc coll)
        db-name (-db-name coll)
        coll-name (-name coll)
        result (-> rpc .-mongo (.aggregate db-name coll-name pipeline))]
    (if (sequential? result) (vec result) [])))

;; ============================================================================
;; Index Operations
;; ============================================================================

(defn create-index!
  "Create an index on the collection.

   Args:
     coll - Collection instance
     keys - Index keys as either:
            - String for single field ascending: \"name\"
            - Vector of [field direction] pairs: [[\"name\" 1] [\"age\" -1]]
     opts - Optional configuration:
            :unique - Create unique index
            :sparse - Create sparse index
            :name - Index name
            :background - Create in background

   Returns:
     Name of the created index.

   Example:
     (create-index! users \"email\" {:unique true})
     (create-index! users [[\"name\" 1] [\"created-at\" -1]])"
  ([coll keys]
   (create-index! coll keys {}))
  ([coll keys opts]
   (let [rpc (-rpc coll)
         db-name (-db-name coll)
         coll-name (-name coll)
         index-keys (if (string? keys) [[keys 1]] keys)
         result (-> rpc .-mongo (.createIndex db-name coll-name index-keys opts))]
     (if (string? result) result ""))))

(defn drop-index!
  "Drop an index from the collection.

   Args:
     coll - Collection instance
     index-name - Name of the index to drop

   Example:
     (drop-index! users \"email_1\")"
  [coll index-name]
  (let [rpc (-rpc coll)
        db-name (-db-name coll)
        coll-name (-name coll)]
    (-> rpc .-mongo (.dropIndex db-name coll-name index-name))
    nil))

(defn drop!
  "Drop the collection.

   Example:
     (drop! temp-collection)"
  [coll]
  (let [rpc (-rpc coll)
        db-name (-db-name coll)
        coll-name (-name coll)]
    (-> rpc .-mongo (.dropCollection db-name coll-name))
    nil))

;; ============================================================================
;; Convenience Functions
;; ============================================================================

(defn find-by-id
  "Find a document by its _id.

   Args:
     coll - Collection instance
     id - Document _id

   Returns:
     Document map, or nil if not found.

   Example:
     (find-by-id users \"abc123\")"
  [coll id]
  (find-one coll {:_id id}))

(defn delete-by-id!
  "Delete a document by its _id.

   Args:
     coll - Collection instance
     id - Document _id

   Returns:
     DeleteResult map.

   Example:
     (delete-by-id! users \"abc123\")"
  [coll id]
  (delete-one! coll {:_id id}))

(defn update-by-id!
  "Update a document by its _id.

   Args:
     coll - Collection instance
     id - Document _id
     update - Update operations map

   Returns:
     UpdateResult map.

   Example:
     (update-by-id! users \"abc123\" {:$set {:status \"active\"}})"
  [coll id update]
  (update-one! coll {:_id id} update))

File: sdks/python/src/mongo_do/collection.py
"""
Collection - MongoDB collection operations.

Provides a PyMongo-compatible Collection interface with async CRUD
operations backed by RPC calls.
"""

from __future__ import annotations

import uuid
from typing import TYPE_CHECKING, Any, Generic, TypeVar

from .cursor import Cursor
from .types import (
    DeleteResult,
    DuplicateKeyError,
    InsertManyResult,
    InsertOneResult,
    UpdateResult,
    WriteError,
)

if TYPE_CHECKING:
    from rpc_do import RpcClient

    from .database import Database
    from .types import Filter, Projection, Update

T = TypeVar("T", bound=dict[str, Any])

__all__ = ["Collection"]


class Collection(Generic[T]):
    """
    MongoDB collection with async CRUD operations.

    Provides a PyMongo-compatible API for interacting with a MongoDB
    collection over RPC. All operations are async.

    Example:
        users = db["users"]

        # Insert
        result = await users.insert_one({"name": "Alice"})
        print(result.inserted_id)

        # Find
        user = await users.find_one({"name": "Alice"})
        async for user in users.find({"status": "active"}):
            print(user)

        # Update
        await users.update_one({"name": "Alice"}, {"$set": {"status": "vip"}})

        # Delete
        await users.delete_one({"name": "Alice"})
    """

    __slots__ = ("_rpc", "_database", "_name", "_full_name")

    def __init__(
        self,
        rpc: RpcClient,
        database: Database,
        name: str,
    ) -> None:
        """
        Initialize a collection.

        Args:
            rpc: The RPC client for making calls.
            database: Parent database instance.
            name: Collection name.
        """
        self._rpc = rpc
        self._database = database
        self._name = name
        self._full_name = f"{database.name}.{name}"

    @property
    def name(self) -> str:
        """Get the collection name."""
        return self._name

    @property
    def full_name(self) -> str:
        """Get the full collection name (database.collection)."""
        return self._full_name

    @property
    def database(self) -> Database:
        """Get the parent database."""
        return self._database

    def _generate_id(self) -> str:
        """Generate a unique document ID."""
        return str(uuid.uuid4())

    async def insert_one(self, document: T) -> InsertOneResult:
        """
        Insert a single document.

        Args:
            document: The document to insert.

        Returns:
            InsertOneResult with the inserted ID.

        Raises:
            DuplicateKeyError: If a document with the same _id exists.
            WriteError: If the insert fails.
        """
        # Generate _id if not provided
        doc = dict(document)
        if "_id" not in doc:
            doc["_id"] = self._generate_id()

        try:
            result = await self._rpc.mongo.insertOne(
                self._database.name,
                self._name,
                doc,
            )

            # Handle RPC result
            if isinstance(result, dict):
                if result.get("error"):
                    error_msg = result.get("message", "Insert failed")
                    if "duplicate" in error_msg.lower() or "E11000" in error_msg:
                        raise DuplicateKeyError(error_msg)
                    raise WriteError(error_msg)
                return InsertOneResult(
                    inserted_id=result.get("insertedId", doc["_id"]),
                    acknowledged=result.get("acknowledged", True),
                )

            return InsertOneResult(inserted_id=doc["_id"])
        except DuplicateKeyError:
            raise
        except WriteError:
            raise
        except Exception as e:
            raise WriteError(str(e)) from e

    async def insert_many(
        self,
        documents: list[T],
        ordered: bool = True,
    ) -> InsertManyResult:
        """
        Insert multiple documents.

        Args:
            documents: List of documents to insert.
            ordered: If True, stop on first error. If False, continue.

        Returns:
            InsertManyResult with the inserted IDs.

        Raises:
            WriteError: If the insert fails.
        """
        # Generate _ids for documents without them
        docs = []
        for document in documents:
            doc = dict(document)
            if "_id" not in doc:
                doc["_id"] = self._generate_id()
            docs.append(doc)

        try:
            result = await self._rpc.mongo.insertMany(
                self._database.name,
                self._name,
                docs,
                {"ordered": ordered},
            )

            if isinstance(result, dict):
                if result.get("error"):
                    raise WriteError(result.get("message", "Insert failed"))
                return InsertManyResult(
                    inserted_ids=result.get("insertedIds", [doc["_id"] for doc in docs]),
                    acknowledged=result.get("acknowledged", True),
                )

            return InsertManyResult(inserted_ids=[doc["_id"] for doc in docs])
        except WriteError:
            raise
        except Exception as e:
            raise WriteError(str(e)) from e

    async def find_one(
        self,
        filter: Filter | None = None,
        projection: Projection = None,
    ) -> T | None:
        """
        Find a single document.

        Args:
            filter: Query filter.
            projection: Fields to include/exclude.

        Returns:
            The matching document, or None if not found.
        """
        options: dict[str, Any] = {}
        if projection:
            if isinstance(projection, list):
                options["projection"] = {field: 1 for field in projection}
            else:
                options["projection"] = dict(projection)

        result = await self._rpc.mongo.findOne(
            self._database.name,
            self._name,
            filter or {},
            options,
        )

        if result is None or (isinstance(result, dict) and result.get("error")):
            return None

        return result  # type: ignore

    def find(
        self,
        filter: Filter | None = None,
        projection: Projection = None,
    ) -> Cursor[T]:
        """
        Find documents matching the filter.

        Args:
            filter: Query filter.
            projection: Fields to include/exclude.

        Returns:
            Cursor for iterating over results.

        Example:
            async for doc in collection.find({"status": "active"}):
                print(doc)

            # With chaining
            cursor = collection.find({}).sort("name").limit(10)
            docs = await cursor.to_list()
        """
        return Cursor[T](
            self._rpc,
            self._database.name,
            self._name,
            filter,
            projection,
        )

    async def update_one(
        self,
        filter: Filter,
        update: Update,
        upsert: bool = False,
    ) -> UpdateResult:
        """
        Update a single document.

        Args:
            filter: Query filter to match the document.
            update: Update operations ($set, $unset, $inc, etc.).
            upsert: If True, insert if no document matches.

        Returns:
            UpdateResult with match/modify counts.

        Raises:
            WriteError: If the update fails.
        """
        try:
            result = await self._rpc.mongo.updateOne(
                self._database.name,
                self._name,
                filter,
                dict(update),
                {"upsert": upsert},
            )

            if isinstance(result, dict):
                if result.get("error"):
                    raise WriteError(result.get("message", "Update failed"))
                return UpdateResult(
                    matched_count=result.get("matchedCount", 0),
                    modified_count=result.get("modifiedCount", 0),
                    upserted_id=result.get("upsertedId"),
                    acknowledged=result.get("acknowledged", True),
                )

            return UpdateResult()
        except WriteError:
            raise
        except Exception as e:
            raise WriteError(str(e)) from e

    async def update_many(
        self,
        filter: Filter,
        update: Update,
        upsert: bool = False,
    ) -> UpdateResult:
        """
        Update multiple documents.

        Args:
            filter: Query filter to match documents.
            update: Update operations ($set, $unset, $inc, etc.).
            upsert: If True, insert if no document matches.

        Returns:
            UpdateResult with match/modify counts.

        Raises:
            WriteError: If the update fails.
        """
        try:
            result = await self._rpc.mongo.updateMany(
                self._database.name,
                self._name,
                filter,
                dict(update),
                {"upsert": upsert},
            )

            if isinstance(result, dict):
                if result.get("error"):
                    raise WriteError(result.get("message", "Update failed"))
                return UpdateResult(
                    matched_count=result.get("matchedCount", 0),
                    modified_count=result.get("modifiedCount", 0),
                    upserted_id=result.get("upsertedId"),
                    acknowledged=result.get("acknowledged", True),
                )

            return UpdateResult()
        except WriteError:
            raise
        except Exception as e:
            raise WriteError(str(e)) from e

    async def replace_one(
        self,
        filter: Filter,
        replacement: T,
        upsert: bool = False,
    ) -> UpdateResult:
        """
        Replace a single document.

        Args:
            filter: Query filter to match the document.
            replacement: The replacement document.
            upsert: If True, insert if no document matches.

        Returns:
            UpdateResult with match/modify counts.

        Raises:
            WriteError: If the replace fails.
        """
        try:
            result = await self._rpc.mongo.replaceOne(
                self._database.name,
                self._name,
                filter,
                dict(replacement),
                {"upsert": upsert},
            )

            if isinstance(result, dict):
                if result.get("error"):
                    raise WriteError(result.get("message", "Replace failed"))
                return UpdateResult(
                    matched_count=result.get("matchedCount", 0),
                    modified_count=result.get("modifiedCount", 0),
                    upserted_id=result.get("upsertedId"),
                    acknowledged=result.get("acknowledged", True),
                )

            return UpdateResult()
        except WriteError:
            raise
        except Exception as e:
            raise WriteError(str(e)) from e

    async def delete_one(self, filter: Filter) -> DeleteResult:
        """
        Delete a single document.

        Args:
            filter: Query filter to match the document.

        Returns:
            DeleteResult with the deleted count.

        Raises:
            WriteError: If the delete fails.
        """
        try:
            result = await self._rpc.mongo.deleteOne(
                self._database.name,
                self._name,
                filter,
            )

            if isinstance(result, dict):
                if result.get("error"):
                    raise WriteError(result.get("message", "Delete failed"))
                return DeleteResult(
                    deleted_count=result.get("deletedCount", 0),
                    acknowledged=result.get("acknowledged", True),
                )

            return DeleteResult()
        except WriteError:
            raise
        except Exception as e:
            raise WriteError(str(e)) from e

    async def delete_many(self, filter: Filter) -> DeleteResult:
        """
        Delete multiple documents.

        Args:
            filter: Query filter to match documents.

        Returns:
            DeleteResult with the deleted count.

        Raises:
            WriteError: If the delete fails.
        """
        try:
            result = await self._rpc.mongo.deleteMany(
                self._database.name,
                self._name,
                filter,
            )

            if isinstance(result, dict):
                if result.get("error"):
                    raise WriteError(result.get("message", "Delete failed"))
                return DeleteResult(
                    deleted_count=result.get("deletedCount", 0),
                    acknowledged=result.get("acknowledged", True),
                )

            return DeleteResult()
        except WriteError:
            raise
        except Exception as e:
            raise WriteError(str(e)) from e

    async def count_documents(self, filter: Filter | None = None) -> int:
        """
        Count documents matching the filter.

        Args:
            filter: Query filter.

        Returns:
            Number of matching documents.
        """
        result = await self._rpc.mongo.countDocuments(
            self._database.name,
            self._name,
            filter or {},
        )
        return result if isinstance(result, int) else 0

    async def estimated_document_count(self) -> int:
        """
        Get an estimated count of documents in the collection.

        This is faster than count_documents() but may not be accurate.

        Returns:
            Estimated number of documents.
        """
        result = await self._rpc.mongo.estimatedDocumentCount(
            self._database.name,
            self._name,
        )
        return result if isinstance(result, int) else 0

    async def distinct(
        self,
        key: str,
        filter: Filter | None = None,
    ) -> list[Any]:
        """
        Get distinct values for a field.

        Args:
            key: Field name to get distinct values for.
            filter: Query filter.

        Returns:
            List of distinct values.
        """
        result = await self._rpc.mongo.distinct(
            self._database.name,
            self._name,
            key,
            filter or {},
        )
        return result if isinstance(result, list) else []

    async def aggregate(self, pipeline: list[dict[str, Any]]) -> list[dict[str, Any]]:
        """
        Run an aggregation pipeline.

        Args:
            pipeline: List of aggregation stages.

        Returns:
            List of aggregation results.
        """
        result = await self._rpc.mongo.aggregate(
            self._database.name,
            self._name,
            pipeline,
        )
        return result if isinstance(result, list) else []

    async def create_index(
        self,
        keys: list[tuple[str, int]] | str,
        **kwargs: Any,
    ) -> str:
        """
        Create an index on the collection.

        Args:
            keys: Index keys as list of (field, direction) tuples,
                  or a single field name.
            **kwargs: Additional index options (unique, sparse, etc.).

        Returns:
            Name of the created index.
        """
        if isinstance(keys, str):
            keys = [(keys, 1)]

        result = await self._rpc.mongo.createIndex(
            self._database.name,
            self._name,
            keys,
            kwargs,
        )
        return result if isinstance(result, str) else ""

    async def drop_index(self, index_name: str) -> None:
        """
        Drop an index from the collection.

        Args:
            index_name: Name of the index to drop.
        """
        await self._rpc.mongo.dropIndex(
            self._database.name,
            self._name,
            index_name,
        )

    async def drop(self) -> None:
        """Drop the collection."""
        await self._rpc.mongo.dropCollection(
            self._database.name,
            self._name,
        )

    def __repr__(self) -> str:
        return f"Collection({self._full_name!r})"
